hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -951,11 +955,12 @@ def draw(self, strategy: "SearchStrategy[Ex]", label: Optional[int] = None) -> "
                 return strategy.do_draw(self)
             else:
                 assert start_time is not None
-                strategy.validate()
-                try:
-                    return strategy.do_draw(self)
-                finally:
-                    self.draw_times.append(time.perf_counter() - start_time)
+                with ensure_free_stackframes():
```

Comment:
I'm pretty sure that we _shouldn't_ do this here, because it's only ever valid to draw when inside a test context, and so it would be redundant with the call in `engine.py` - and likely unsafe, too.

Reply:
I'm having some difficulty in consistently reproducing this morning, but it looks like something is needed in the vicinity. Would it be better to move it up a stackframe, to
https://github.com/HypothesisWorks/hypothesis/blob/981d01e7bd015058e3443bf28afd692411bc3bb3/hypothesis-python/src/hypothesis/control.py#L87
?

The stack down to this point is

```
  File "/home/jobh/src/hypothesis/hypothesis-python/src/hypothesis/core.py", line 1392, in wrapped_test
    raise the_error_hypothesis_found
  File "/home/jobh/src/hypothesis/hypothesis-python/src/hypothesis/core.py", line 1359, in wrapped_test
    state.run_engine()
  File "/home/jobh/src/hypothesis/hypothesis-python/src/hypothesis/core.py", line 1025, in run_engine
    _raise_to_user(errors_to_report, self.settings, report_lines)
  File "/home/jobh/src/hypothesis/hypothesis-python/src/hypothesis/core.py", line 1061, in _raise_to_user
    raise the_error_hypothesis_found
  File "/home/jobh/src/hypothesis/hypothesis-python/src/hypothesis/core.py", line 986, in run_engine
    self.execute_once(
  File "/home/jobh/src/hypothesis/hypothesis-python/src/hypothesis/core.py", line 807, in execute_once
    result = self.test_runner(data, run)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jobh/src/hypothesis/hypothesis-python/src/hypothesis/executors.py", line 47, in default_new_style_executor
    return function(data)
           ^^^^^^^^^^^^^^
  File "/home/jobh/src/hypothesis/hypothesis-python/src/hypothesis/core.py", line 769, in run
    a, kw, argslices = context.prep_args_kwargs_from_strategies(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jobh/src/hypothesis/hypothesis-python/src/hypothesis/control.py", line 94, in prep_args_kwargs_from_strategies
    obj = self.data.draw(s)
          ^^^^^^^^^^^^^^^^^
  File "/home/jobh/src/hypothesis/hypothesis-python/src/hypothesis/internal/conjecture/data.py", line 956, in draw
    return strategy.do_draw(self)
           ^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
Huh, I would have been very confident that `self.test_runner(data, run)` went through our context manager in the engine _before_ we got to draw anything from a strategy.  Check whether disabling trimming in [`get_trimmed_traceback()`](https://github.com/HypothesisWorks/hypothesis/blob/981d01e7bd015058e3443bf28afd692411bc3bb3/hypothesis-python/src/hypothesis/internal/escalation.py#L82) changes what you see?  

If not, we probably want to _move_ the current context in the engine to the `execute_once()`, so long as our `.fuzz_one_input` logic also passes through there (which I _think_ it does).

Reply:
Ah, yes, now I see it (after disabling trimming).

It does indeed go through the engine, and it was an artifact of using the (insufficient) artificially lowered stack limit as increment.